### ISSUE TITLE
"tmux new-window \; split-window -h" does not work as expected anymore.

### ISSUE THREAD

**Issue by MadMaverick9** [tmux/tmux#102]
Start tmux with:
```
bash $ tmux
```
Which will start a session with one window and a shell started. In that shell type:
```
bash $ tmux new-window \; split-window
```
Up to 1.9a this would create a new window and split that new window.
In 2.0 and above it creates a new window, but then splits the first window.
```
1441098112.556005 cmdq 0x841e928: new-window
1441098112.556025 cmd_find_index: target none
1441098112.556033       s=$0
1441098112.556040       wl=0 1 w=@0 bash
1441098112.556047       wp=%0
1441098112.556053       idx=none
1441098112.556134 @1 alerts flags added 0x2
1441098112.556146 @1 alerts timer reset 0
1441098112.556167 spawn: /bin/bash -- 
1441098112.558284 cmdq 0x841e928: split-window
1441098112.558312 cmd_find_pane: target none
1441098112.558321       s=$0
1441098112.558327       wl=0 1 w=@0 bash
1441098112.558334       wp=%0
1441098112.558341       idx=0
1441098112.558427 spawn: /bin/bash -- 
1441098112.560463 session 0 has 2 references
```
This is an excerpt from the log file and it seems to find (see "`cmd_find_pane:`") the wrong window "`wl=0 1 w=@0 bash`", which should be "`wl=1 1 w=@1 bash`" if I am not mistaken.

When I do `Ctrl-B:` and then type `new-window ; split-window` it works fine. So it seems to have something to do with that when you type "tmux new-w...." in a shell, it is a second client connecting to the session.

That's as far as I got. I am stuck.

Thank you for your help.

**Comment by nicm** [comment]
This is sort of intended. Previously, when you ran things from the shell, tmux would use the active pane. Now it uses the pane where the command was run (that is, where the shell is running). Often they are the same, but not in this case. This is better sometimes, but this is a case where it is not obvious. It would be nice to track a "working pane" through command sequences (so it starts where the command was run and changes with new/neww/splitw/etc), but we currently don't do that.

**Comment by nicm** [comment]
This http://pastebin.ca/raw/3146986 is the general idea and fixes your case, but I am not very comfortable with it. Because of the way the "current session" can only ever be a guess unless the command is run from inside tmux itself (with a key binding or command prompt), there are going to be cases with new-session where you end up running the command in the wrong session. It would be better if instead of a flag it actually saved the session/window/etc but then we need to verify they are still present and valid when we come to use them for the next command. It is not hugely complicated (we already validate the marked pane if needed) but more than this.

**Comment by nicm** [comment]
This does that http://pastebin.ca/raw/3147032 but there are lots of other places than change window or pane, for example select-pane. Not certain this is a good approach.

**Comment by MadMaverick9** [comment]
Thank you very much for your info. Your info and code diffs clear up several questions I had yesterday.
> the "current session" can only ever be a guess

This is one thing that confused me yesterday. Isn't a client always connected to a specific pane and therefor a specific session when issuing a command from the command line (shell) and/or from within tmux? So why the guessing based on activity time?
```
return (timercmp(&s->activity_time, &than->activity_time, >));
```
I guess you answered this already by writing that we need to keep track of a "working pane".

Yet another testcase:
```
bash $ sleep 10 ; tmux new-window \; split-window
```
And if you switch to a different session within those ten seconds you will get unexpected results now. Correct? Unless we keep track of a "working pane" and "working session"? Correct? Your answer to this would definitely help my understanding of the code.

After your answer I think I can close this issue and start tinkering.

Thank you very much.

**Comment by nicm** [comment]
On Tue, Sep 01, 2015 at 09:18:59PM -0700, MadMaverick9 wrote:
>    Thank you very much for your info. Your info and code diffs clear up
>    several questions I had yesterday.
> 
>      the "current session" can only ever be a guess
> 
>    This is one thing that confused me yesterday. Isn't a client always
>    connected to a specific pane and therefor a specific session when issuing
>    a command from the command line (shell) and/or from within tmux? So why
>    the guessing based on activity time?

If you run a tmux client in a shell, we can use the pty to find the
pane. Panes can only be in one window, so we can also find the
window. But windows can be in any number of sessions, so we have no sure
way to know which session should be used. The activity time gives us a
reasonable guess, it should be correct if the user was actually typing
in a session.

This is not the case for the command prompt (or a key binding) where we
know the actual client the user typed into, so we also know the exact
current session, window and pane.

>  return (timercmp(&s->activity_time, &than->activity_time, >));
> 
>    I guess you answered this already by writing that we need to keep track of
>    a "working pane".
> 
>    Yet another testcase:
> 
>  bash $ sleep 10 ; tmux new-window \; split-window
> 
>    And if you switch to a different session within those ten seconds you will
>    get unexpected results now. Correct? Unless we keep track of a "working
>    pane" and "working session"? Correct? Your answer to this would definitely
>    help my understanding of the code.

This is the other case, yes.

From the command prompt, when the state is changed by neww the splitw
command will work out a different window (c->session->curw has changed).

From the shell, neww doesn't change the pty, so the pane we worked out
is the same, so we use the same window, session, client.

There are a few things we can do here:

a) We could leave it as it is; what it currently does it the best guess,
which is what you are asking for when you don't give -t.

b) We could make the command prompt work like the shell, save the state
of the first command and use it for them all (if it stays valid). This
is easy but I think it is not obvious for the user.

c) We could finish what my second diff does and make it so that we start
with the best guess and let each command change it as it goes along (so
neww will change it to the new window, etc). The problem is there are a
quite a lot of places commands can change the state.

**Comment by MadMaverick9** [comment]
First - Why did you remove a crucial piece of code with this commit  https://github.com/tmux/tmux/commit/6a539c00dfe52f588376d997a75e6c7540db5e14 ?? Why crucial? Because it held the solution for this bug.

Coincidence? I don't know.

When tmux is run as a client in a shell, we know exactly what session it is running in. The information is in the TMUX environment variable. No guessing needed. That crucial piece of code mentioned above actually parsed that environment variable, but the session id was never used further in the code.

Why not? I don't know.

With the "attached" patch, tmux now works again like it used to before 2.0. Sure makes me happy.

Github does not allow to attach a patch file to a comment, only an image. Why? I don't know.

I will continue using my patched tmux and provide updates here if my patch breaks anything else. But it's looking good so far.

```
Index: client.c
===================================================================
--- client.c	(revision 11)
+++ client.c	(working copy)
@@ -55,7 +55,7 @@
 __dead void	client_exec(const char *);
 int		client_get_lock(char *);
 int		client_connect(struct event_base *, char *, int);
-void		client_send_identify(void);
+void		client_send_identify(int env_session_id);
 int		client_write_one(enum msgtype, int, const void *, size_t);
 int		client_write_server(enum msgtype, const void *, size_t);
 void		client_update_event(void);
@@ -205,7 +205,7 @@
 
 /* Client main loop. */
 int
-client_main(struct event_base *base, int argc, char **argv, int flags)
+client_main(struct event_base *base, int argc, char **argv, int flags, int env_session_id)
 {
 	struct cmd		*cmd;
 	struct cmd_list		*cmdlist;
@@ -304,7 +304,7 @@
 	}
 
 	/* Send identify messages. */
-	client_send_identify();
+	client_send_identify(env_session_id);
 
 	/* Send first command. */
 	if (msg == MSG_COMMAND) {
@@ -359,7 +359,7 @@
 
 /* Send identify messages to server. */
 void
-client_send_identify(void)
+client_send_identify(int env_session_id)
 {
 	const char	 *s;
 	char		**ss;
@@ -369,6 +369,10 @@
 
 	client_write_one(MSG_IDENTIFY_FLAGS, -1, &flags, sizeof flags);
 
+	if (env_session_id >= 0) {
+		client_write_one(MSG_IDENTIFY_ENV_SESSION_ID, -1, &env_session_id, sizeof env_session_id);
+	}
+
 	if ((s = getenv("TERM")) == NULL)
 		s = "";
 	client_write_one(MSG_IDENTIFY_TERM, -1, s, strlen(s) + 1);
Index: cmd-find.c
===================================================================
--- cmd-find.c	(revision 11)
+++ cmd-find.c	(working copy)
@@ -284,9 +284,14 @@
 {
 	/* If we know the current client, use it. */
 	if (fs->cmdq->client != NULL) {
-		if (fs->cmdq->client->session == NULL)
+		if ((fs->cmdq->client->session == NULL) && (fs->cmdq->client->env_session == NULL)) {
 			return (cmd_find_current_session_with_client(fs));
-		fs->s = fs->cmdq->client->session;
+		}
+		if (fs->cmdq->client->env_session != NULL) {
+			fs->s = fs->cmdq->client->env_session;
+		} else {
+			fs->s = fs->cmdq->client->session;
+		}
 		fs->wl = fs->s->curw;
 		fs->idx = fs->wl->idx;
 		fs->w = fs->wl->window;
Index: server-client.c
===================================================================
--- server-client.c	(revision 11)
+++ server-client.c	(working copy)
@@ -110,6 +110,7 @@
 
 	c->session = NULL;
 	c->last_session = NULL;
+	c->env_session = NULL;
 	c->tty.sx = 80;
 	c->tty.sy = 24;
 
@@ -1015,6 +1016,7 @@
 		case MSG_IDENTIFY_STDIN:
 		case MSG_IDENTIFY_ENVIRON:
 		case MSG_IDENTIFY_CLIENTPID:
+		case MSG_IDENTIFY_ENV_SESSION_ID:
 		case MSG_IDENTIFY_DONE:
 			server_client_msg_identify(c, &imsg);
 			break;
@@ -1149,7 +1151,8 @@
 {
 	const char	*data;
 	size_t	 	 datalen;
-	int		 flags;
+	int		 flags, env_session_id;
+	struct session  *env_session;
 
 	if (c->flags & CLIENT_IDENTIFIED)
 		fatalx("out-of-order identify message");
@@ -1195,6 +1198,15 @@
 			fatalx("bad MSG_IDENTIFY_CLIENTPID size");
 		memcpy(&c->pid, data, sizeof c->pid);
 		break;
+	case MSG_IDENTIFY_ENV_SESSION_ID:
+		if (datalen != sizeof env_session_id)
+			fatalx("bad MSG_IDENTIFY_ENV_SESSION_ID size");
+		memcpy(&env_session_id, data, sizeof env_session_id);
+		env_session = session_find_by_id(env_session_id);
+		if (env_session != NULL) {
+			c->env_session = env_session;
+		}
+		break;
 	default:
 		break;
 	}
Index: tmux.c
===================================================================
--- tmux.c	(revision 11)
+++ tmux.c	(working copy)
@@ -201,6 +201,8 @@
 {
 	char	*s, *path, *label, **var, tmp[PATH_MAX];
 	int	 opt, flags, keys;
+	char	 env_socket_path[256];
+	int	 env_pid, env_session_id;
 
 #if defined(DEBUG) && defined(__OpenBSD__)
 	malloc_options = (char *) "AFGJPX";
@@ -318,6 +320,14 @@
 		options_set_number(&global_w_options, "mode-keys", keys);
 	}
 
+        /* Get path from environment. */
+	s = getenv("TMUX");
+	if (s == NULL || sscanf(s, "%255[^,],%lld,%d", env_socket_path, &env_pid, &env_session_id) != 3) {
+		env_socket_path[0] = '\0';
+		env_pid = -1;
+		env_session_id = -1;
+	}
+
 	/*
 	 * Figure out the socket path. If specified on the command-line with -S
 	 * or -L, use it, otherwise try $TMUX or assume -L default.
@@ -325,16 +335,11 @@
 	if (path == NULL) {
 		/* If no -L, use the environment. */
 		if (label == NULL) {
-			s = getenv("TMUX");
-			if (s != NULL) {
-				path = xstrdup(s);
-				path[strcspn (path, ",")] = '\0';
-				if (*path == '\0') {
-					free(path);
-					label = xstrdup("default");
-				}
-			} else
+			if (env_socket_path[0] != '\0') {
+				path = xstrdup(env_socket_path);
+			} else {
 				label = xstrdup("default");
+			}
 		}
 
 		/* -L or default set. */
@@ -348,8 +353,7 @@
 	}
 	free(label);
 
-	if (strlcpy(socket_path, path, sizeof socket_path) >=
-	    sizeof socket_path) {
+	if (strlcpy(socket_path, path, sizeof socket_path) >= sizeof socket_path) {
 		fprintf(stderr, "socket path too long: %s\n", path);
 		exit(1);
 	}
@@ -361,5 +365,5 @@
 #endif
 
 	/* Pass control to the client. */
-	exit(client_main(osdep_event_init(), argc, argv, flags));
+	exit(client_main(osdep_event_init(), argc, argv, flags, env_session_id));
 }
Index: tmux.h
===================================================================
--- tmux.h	(revision 11)
+++ tmux.h	(working copy)
@@ -401,6 +401,7 @@
 	MSG_IDENTIFY_ENVIRON,
 	MSG_IDENTIFY_DONE,
 	MSG_IDENTIFY_CLIENTPID,
+	MSG_IDENTIFY_ENV_SESSION_ID,
 
 	MSG_COMMAND = 200,
 	MSG_DETACH,
@@ -1247,6 +1248,7 @@
 
 	struct session	*session;
 	struct session	*last_session;
+	struct session  *env_session;
 
 	int		 wlmouse;
 
@@ -1702,7 +1704,7 @@
 void	cmd_wait_for_flush(void);
 
 /* client.c */
-int	client_main(struct event_base *, int, char **, int);
+int	client_main(struct event_base *, int, char **, int, int);
 
 /* key-bindings.c */
 RB_PROTOTYPE(key_bindings, key_binding, entry, key_bindings_cmp);
```

**Comment by nicm** [comment]
Good idea, but I think you should verify that the window is still in the session (check that the client tty belongs to a pane whose window is in the session). $TMUX can easily be stale which is why we have moved away from using it.

Also I think you should set c->env_session to NULL when the session is killed, I'm not sure there is actually a way to make it be dereferenced but it would be better not to leave stale pointers around anyway.

**Comment by nicm** [comment]
Also, I think your diff is using env_session too early. You have made it use not only the env_session but the /current window/ of env_session. We don't want to do this, because we KNOW the window and pane - we can find it out by looking at the pty.

Instead, cmd_find_current_session_with_client should use the env_session if present (and if wp is still in env_session) instead of calling cmd_find_best_session_with_window.

**Comment by nicm** [comment]
To explain a bit - c->env_session cannot be a replacement for c->session: c->session set means that the command was run from inside tmux, we can be completely certain that the user wants that session and its current window and current pane. When c->session is not set, we have to decide what to use. If the client tty matches the tty of a pane, we can be completely certain that the user ran tmux inside that pane, so that should be the target pane. If $TMUX is set, we can be mostly certain that is the session we want and if the session exists and the client tty pane is inside it, we can use that session. If the client tty pane isn't inside it, clearly the pane has moved so we need to go back to guessing.

However, this then won't solve your original problem, because the problem is not that we have to guess the session (although that is not ideal), the problem is that both commands (neww and splitw) use the same context to decide the target pane and they both use the pane where the command is run, when in fact it would be better if neww used that pane, and then splitw used the current pane (which is what my diffs above attempt to do)

**Comment by nicm** [comment]
"splitw used the current pane" <-- no, not the current pane, the neww _target_ pane.

**Comment by nicm** [comment]
And one last comment on the diff, you don't need to parse TMUX in the client and add a new IDENTIFY type - you should be able to get it from c->environ in the server instead.

**Comment by MadMaverick9** [comment]
ugh - you're right. :)

During my daily use of tmux, I move windows around, but never move em to a different session. So in my personal case TMUX is valid at all times. But you have a good point; it will go stale quickly in use cases where users move windows to other sessions or link windows to other sessions.

Your continued patience and input is much appreciated.

Back to the drawing board.

**Comment by MadMaverick9** [comment]
Ok - how about this?
```
Index: cmd-find.c
The idea for this patch comes from:
tmux-1.9a.tar.gz: ./tmux-1.9a/cmd.c (Line 1129)
Patch below restores previous behavior.
===================================================================
--- cmd-find.c  (revision 13)
+++ cmd-find.c  (working copy)
@@ -272,6 +272,10 @@
        /* Find the best session and winlink. */
        if (cmd_find_best_session_with_window(fs) != 0)
                return (-1);
+       fs->wl = fs->s->curw;
+       fs->idx = fs->wl->idx;
+       fs->w = fs->wl->window;
+       fs->wp = fs->w->active;
        return (0);
 }
 
```

**Comment by nicm** [comment]
We don't want to restore the previous behaviour, we want to continue using the pty for the first command and use the new window for the second. Your change would use the current window for all commands.

**Comment by nicm** [comment]
Take a look at this which is a simpler version of your idea to use $TMUX, and uses it to find the session when we have a choice (either when we definitely know the window, or we definitely don't know the window). Note that this doesn't have any impact on the original issue.

Also of course this also changes behaviour, if you have a window linked into two sessions, it will always prefer the original session if it still exists even if you actually typed the command into a different session :-).

```
? tmux
? tmux-client-29386.log
? tmux-out-31689.log
? tmux-server-31689.log
Index: cmd-find.c
===================================================================
RCS file: /cvs/src/usr.bin/tmux/cmd-find.c,v
retrieving revision 1.13
diff -u -p -r1.13 cmd-find.c
--- cmd-find.c	14 Sep 2015 13:22:02 -0000	1.13
+++ cmd-find.c	22 Sep 2015 11:27:23 -0000
@@ -23,6 +23,7 @@
 #include <stdlib.h>
 #include <string.h>
 #include <paths.h>
+#include <unistd.h>
 
 #include "tmux.h"
 
@@ -51,6 +52,7 @@ struct cmd_find_state {
 	int			 idx;
 };
 
+struct session	*cmd_find_try_TMUX(struct client *, struct window *);
 int		 cmd_find_client_better(struct client *, struct client *);
 struct client	*cmd_find_best_client(struct client **, u_int);
 int		 cmd_find_session_better(struct session *, struct session *,
@@ -109,6 +111,33 @@ const char *cmd_find_pane_table[][2] = {
 	{ NULL, NULL }
 };
 
+/* Get session from TMUX if present. */
+struct session *
+cmd_find_try_TMUX(struct client *c, struct window *w)
+{
+	struct environ_entry	*envent;
+	char			 tmp[256];
+	long long		 pid;
+	u_int			 session;
+	struct session		*s;
+
+	envent = environ_find(&c->environ, "TMUX");
+	if (envent == NULL)
+		return (NULL);
+
+	if (sscanf(envent->value, "%255[^,],%lld,%d", tmp, &pid, &session) != 3)
+		return (NULL);
+	if (pid != getpid())
+		return (NULL);
+	log_debug("client %d TMUX is %s (session @%u)", c->ibuf.fd,
+	    envent->value, session);
+
+	s = session_find_by_id(session);
+	if (s == NULL || (w != NULL && !session_has(s, w)))
+		return (NULL);
+	return (s);
+}
+
 /* Is this client better? */
 int
 cmd_find_client_better(struct client *c, struct client *than)
@@ -192,6 +221,12 @@ cmd_find_best_session_with_window(struct
 	u_int		  ssize;
 	struct session	 *s;
 
+	if (fs->cmdq->client != NULL) {
+		fs->s = cmd_find_try_TMUX(fs->cmdq->client, fs->w);
+		if (fs->s != NULL)
+			return (cmd_find_best_winlink_with_window(fs));
+	}
+
 	ssize = 0;
 	RB_FOREACH(s, sessions, &sessions) {
 		if (!session_has(s, fs->w))
@@ -256,7 +291,10 @@ cmd_find_current_session_with_client(str
 
 	/* Not running in a pane. We know nothing. Find the best session. */
 	if (wp == NULL) {
-		fs->s = cmd_find_best_session(NULL, 0, fs->flags);
+		if (fs->cmdq->client != NULL)
+			fs->s = cmd_find_try_TMUX(fs->cmdq->client, NULL);
+		if (fs->s == NULL)
+			fs->s = cmd_find_best_session(NULL, 0, fs->flags);
 		if (fs->s == NULL)
 			return (-1);
 		fs->wl = fs->s->curw;

**Comment by nicm** [comment]
Hmm, I don't know if the rules we currently use are actually the most sensible, perhaps they are not so good for interactive use. But I like the fact that if I run something in a pane that later runs tmux, it will use that pane (from the pty) rather than picking whatever pane I happen to be working in.

**Comment by dbingham** [comment]
So, I definitely understand the limitations you are describing here and I don't pretend to have a clue how to resolve the issue (I'll leave that to you guys as experts).  My question is, as a user, is there a mechanism that I can use to achieve a similar result as before?  I'm not afraid of doing some shell scripting to achieve it.  For example, in pseudo-code:

```
newWinId = new_window
split_window -t ${newWinId} -h
```

I know the commands don't work that way, just trying to illustrate the idea.  Basically, I want a way for the first action to tell me the ID of the pane/window/session of its final state so that I can remove the ambiguity upon invoking the second command.

Maybe something like this (though, I admit to struggling to remember the whole tmux man page, so maybe there is a better way):
```
tmux neww "echo ${TMUX_PANE} > /tmp/targetID && bash"
tmux splits -h -t $(cat /tmp/targetID)
```
(I tried this and it doesn't work... the redirection to file didn't happen so, obvious, the cat didn't work and the command failed)

Basically, I have bash scripts to create new windows, sessions, etc in various configurations and I'm trying to figure out how to update them given the new model.  Any help in this regard would be much appreciated.

**Comment by nicm** [comment]
You can use -P to make neww or splitw print what it created and use that with -t in subsequent commands.

**Comment by dbingham** [comment]
Well, dang.  How is it I've read the `new-window` entry in the man page about 50 times and missed that every time.  `TARGET=$(tmux neww -P) && tmux splitw -h -t ${TARGET}` works perfectly.  Thanks!

**Comment by aubergene** [comment]
I don't know much about tmux, but I use this script to open multiple windows and it's no longer working with version 2.1. Would it be possible to let me know what I need to change to get it to function again. Thanks

https://gist.github.com/dmytro/3984680

**Comment by jody-frankowski** [comment]
@aubergene hey, here is a tested gist that uses the new tmux behavior: https://gist.github.com/jody-frankowski/76e6ad00eaf810d2e7ac
It works for zsh or bash.

**Comment by aubergene** [comment]
@jody-frankowski Thank you very much for this! Things were super busy and it took me until today to fix my script with your suggestion.

**Comment by unphased** [comment]
I noticed with Tmux 2.2 master branch that `tmux` command calls made from inside a tmux pane will act on the session that the pane is associated or was created in. I can also see that `$TMUX` has a string inside that corresponds to the session name. 

This is cool I guess, but it generally breaks a large number of the configuration I have created. You see, sometimes it is very very useful for (e.g. vim or anything else) to trigger a tmux action. But, if I happen to be in a new fresh session attached to a given pane, then when I do this, now the command takes effect in the original session which is looking at another window.

One might argue that this isnt part of the design of the feature and what have you but it is simply unnatural, and moreover has worked perfectly for 3+ years. Could we perhaps have a flag to `tmux` to force back the original behavior (e.g. from session A if we run a `tmux` command we apply the action to session A, instead of applying the action to the session which originally created the pane)?
